spec/project_spec.rb

```diff
@@ -250,6 +250,52 @@ class InvalidErrorHandler; end
+      expect(Optimizely::Helpers::Validator.inputs_valid?(false, spy_logger, Logger::ERROR)).to eq(false)
+    end
+
+    it 'should log and return false when non string value inputs are passed' do
+      expect(Optimizely::Helpers::Validator.inputs_valid?({user_id: nil}, spy_logger, Logger::ERROR)).to eq(false)
+      expect(Optimizely::Helpers::Validator.inputs_valid?({user_id: ''}, spy_logger, Logger::ERROR)).to eq(false)
+      expect(Optimizely::Helpers::Validator.inputs_valid?({user_id: []}, spy_logger, Logger::ERROR)).to eq(false)
+      expect(Optimizely::Helpers::Validator.inputs_valid?({user_id: {}}, spy_logger, Logger::ERROR)).to eq(false)
+      expect(Optimizely::Helpers::Validator.inputs_valid?({user_id: 2}, spy_logger, Logger::ERROR)).to eq(false)
+      expect(Optimizely::Helpers::Validator.inputs_valid?({user_id: 2.0}, spy_logger, Logger::ERROR)).to eq(false)
+      expect(Optimizely::Helpers::Validator.inputs_valid?({user_id: true}, spy_logger, Logger::ERROR)).to eq(false)
+      expect(Optimizely::Helpers::Validator.inputs_valid?({user_id: false}, spy_logger, Logger::ERROR)).to eq(false)
+      expect(spy_logger).to have_received(:log).with(Logger::ERROR, 'User ID is invalid').exactly(8).times
+    end
+
+    it 'should log and return false when multiple non string value inputs are passed' do
+      expect(Optimizely::Helpers::Validator.inputs_valid?({user_id: '', experiment_key: true}, spy_logger, Logger::ERROR)).to eq(false)
+      expect(Optimizely::Helpers::Validator.inputs_valid?({user_id: [], variation_key: 2.0}, spy_logger, Logger::ERROR)).to eq(false)
+      expect(spy_logger).to have_received(:log).twice.with(Logger::ERROR, 'User ID is invalid')
+      expect(spy_logger).to have_received(:log).once.with(Logger::ERROR, 'Experiment Key is invalid')
+      expect(spy_logger).to have_received(:log).once.with(Logger::ERROR, 'Variation Key is invalid')
+    end
+
+    it 'should return true when valid input values are passed' do
+      expect(Optimizely::Helpers::Validator.inputs_valid?({user_id: '2'}, spy_logger, Logger::ERROR)).to eq(true)
+      expect(Optimizely::Helpers::Validator.inputs_valid?({
+                                                            user_id: 'test_user',
+                                                            experiment_key: 'test_experiment',
+                                                            variation_key: 'test_variation'
+                                                          }, spy_logger, Logger::ERROR)).to eq(true)
+    end
+
+    it 'should should not log when logger or level are nil' do
+      expect(Optimizely::Helpers::Validator.inputs_valid?({user_id: nil}, nil, Logger::ERROR)).to eq(false)
+      expect(Optimizely::Helpers::Validator.inputs_valid?({user_id: nil}, spy_logger, nil)).to eq(false)
+      expect(spy_logger).not_to have_received(:log).with(Logger::ERROR, 'User ID is invalid')
```

Comment:
Remove the with clause here. 
We want to assert that spy_logger does not call :log with any arguments. This line asserts that :log is not called with the **with** clause

Reply:
There is no other logger inside this method. With will confirm assertion.